### ISSUE TITLE
Allow inlined extending within included templates.

### ISSUE THREAD

**Issue by genusistimelord** [djc/askama#572]
So here is the issue, in the current Jinja i can use a card.html which include something like this

```
<div class="uk-card uk-card-default">
    <div class="uk-card-header">
        <span class="uk-card-title">
            {% block card_title %}{% endblock card_title %}
        </span>
        {% block card_badge %}{% endblock card_badge %}
    </div>
    <div class="uk-card-body">
        {% block card_body %}{% endblock card_body %}
    </div>
    {% block card_footer %}{% endblock card_footer %}
</div>
```

so then when i go to setup certain things lets say I want to make a Article Card that uses the card.html to extend itself 

```
{% extends "components/card.html" %}

{% block card %}
    {% block card_title %}
            Registration Agreement
    {% endblock card_title %}
    {% block card_body %}
                    <p>This is a random Article</p>
                    <p><b>Last update: August 9, 2018</b></p>
                    <p><b>By continuing you agree to the above privacy policy.</b></p>
    {% endblock card_body %}
{% endblock card %}
```

This would in jinja load the single extension, Place the changes inside it, then send the changed version down the line to be used in the parent. However Askama does not support this. Would it be Possible to add this or to add it as an inline_extend type? As this is useful for reducing HTML that may get modified to change theme looks ETC without having to go thru 1000 files to do so.

**Comment by djc** [comment]
Sorry, it's not really clear to me what you mean by this. Can you do a more complete example of what you're trying to achieve in terms of output and how reuse should facilitate that?

**Comment by genusistimelord** [comment]
I think the major issue is you only allow parents to Extend so in askamas Case it would only allow
<- means to extend -> means to include

base<-forum->category-card

for this to work id need to be able to 
base<-forum->card<-category-card

**Comment by djc** [comment]
Ahhh, so you want includes to support inheritance? I don't see any fundamental reason why that should not work, might just be an oversight in the parser or code generator. Do you get an error message when you try this?

**Comment by genusistimelord** [comment]
yeah inheritance would be nice. Currently when i try this it tells me extend only works for the parent. Though we could probably make it more apparent that it is an inheritance rather than an extend. Maybe by making it {% inherit "components/category-card.html" %} rather than {% extend "components/category-card.html" %}.  If this makes it easier to introduce and fix. Otherwise if it is a easy fix then I would not worry about the rename. =3.

**Comment by djc** [comment]
Not sure it is an easy fix. I won't be renaming the keyword, since `extends` is in wide use in things such as Jinja/Twig as well.

So did you get an error message?

**Comment by genusistimelord** [comment]
Yes the error is

error: extend blocks only allowed at the top level
   --> src\main.rs:141:10
    |
141 | #[derive(Template, Deserialize, Serialize)]

**Comment by djc** [comment]
Okay, I don't remember why that is exactly but it does make this somewhat hairy probably. I'm not sure I'll have time to work on this in the near future, but you're welcome to have a look at the code. The relevant code is in `askama_shared/src/generator.rs` in `handle_include()`.

**Comment by genusistimelord** [comment]
Yeah so i was somewhat looking into this and it looks as if in order to properly do it back-end  each include would need to be loaded as its own thing first generate the end code and the publish it back to the original code. Not entirely 100% sure but I'm having difficulties following it. So if we did it this way it would allow an include to take in a extend without issue and treat it as the parent until it is returned.

**Comment by djc** [comment]
I don't have objections to that on basis of this high-level description, but I still have a hard time imagining the impact on the code generator so we'd have to see what the sticking points are.

**Comment by genusistimelord** [comment]
idk from looking at the code I'm still trying to fully understand how you currently do things but it seems for included stuff it doesn't load anything extend wise currently. I am just unsure about it ATM as I'm not the best at parsers XD. Like in my mind I would think that we would Grab the extend and use it as the parent object and then Grab the object extending to it and place its blocked code into the base. Doing this per each part that is extended. 

Then after we place all the parts together in each extend we would go and do all the final processing. Handling if statements, Macros etc. This to me seems like a multi parse process similar to compilers. I am still unsure though.

The main point of doing this would be to allow users to simplify their HTML code and reuse portions of it thru extends. This makes it easier to maintain or change an entire websites layout. It is not 100% needed as people could still just write it all out manually. I look at this as more of a if you had to do a theme update or have multiple site themes for like a Forum software or something else that this would make doing so much easier to deal with.

**Comment by djc** [comment]
Yeah, it's niche enough that I don't have much time to spend on it. I think there are other mechanisms in Askama that help with code re-use and can probably also do a decent job covering this use case.

**Comment by genusistimelord** [comment]
OK well we can leave this here as a placeholder. Though I was looking at @malyn  block partials stuff which would work in a similar way.
